Update mysql integration

jsoriano opened the pull request:
## What does this PR do?

* Sync mysql integration with beats, add test files.
* Add reference to ECS fields.
* Add dataset and module fields.

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/master/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] If I'm introducing a new feature, I have modified the Kibana version constraint in my package's `manifest.yml` file to point to the latest Elastic stack release (e.g. `^7.13.0`).

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [ ]

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Supersedes #123
-->
-

## Screenshots

<!-- Optional
Add here screenshots presenting:
- Kibana UI forms presenting configuration options exposed by the integration
- dashboards with collected metrics or logs
-->

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-1145/detail/PR-1145/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-1145/detail/PR-1145/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-1145/detail/PR-1145/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-1145/detail/PR-1145/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://integrations_1145.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats


* Build Cause: `Pull request #1145 opened`



* Start Time: 2021-06-17T15:48:17.549+0000


* Duration: 7 min 0 sec


* Commit: 7b954335e0ed59a13624d52887c260cb06376fa8




#### Trends :test_tube:


![Image of Build Times](https://beats-ci.elastic.co/job/Ingest-manager/job/integrations/job/PR-1145/buildTimeGraph/png)



</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->









### Steps errors [![1](https://img.shields.io/badge/1%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-1145/detail/PR-1145/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Check integration: mysql`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-1145/runs/1/steps/394/log/?start=0">here</a></li>
<li>Description: <code>../../build/elastic-package check -v</code></l1>
</ul>
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2021-06-17T15:54:36.172Z] go: downloading github.com/gregjones/httpcache v0.0.0-20180305231024-9cad4c3443a7
[2021-06-17T15:54:36.172Z] go: downloading github.com/go-logr/logr v0.2.0
[2021-06-17T15:54:36.172Z] go: downloading k8s.io/kubectl v0.20.1
[2021-06-17T15:54:36.172Z] go: downloading sigs.k8s.io/structured-merge-diff/v4 v4.0.2
[2021-06-17T15:54:36.172Z] go: downloading github.com/google/go-cmp v0.5.2
[2021-06-17T15:54:36.172Z] go: downloading github.com/go-git/gcfg v1.5.0
[2021-06-17T15:54:36.172Z] go: downloading github.com/peterbourgon/diskv v2.0.1+incompatible
[2021-06-17T15:54:36.172Z] go: downloading github.com/davecgh/go-spew v1.1.1
[2021-06-17T15:54:36.172Z] go: downloading github.com/kylelemons/godebug v1.1.0
[2021-06-17T15:54:36.172Z] go: downloading github.com/kevinburke/ssh_config v0.0.0-20190725054713-01f96b0aa0cd
[2021-06-17T15:54:36.172Z] go: downloading github.com/xanzy/ssh-agent v0.2.1
[2021-06-17T15:54:36.172Z] go: downloading github.com/google/btree v1.0.0
[2021-06-17T15:54:36.172Z] go: downloading github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
[2021-06-17T15:54:36.172Z] go: downloading k8s.io/kube-openapi v0.0.0-20201113171705-d219536bb9fd
[2021-06-17T15:54:36.172Z] go: downloading github.com/go-openapi/strfmt v0.19.6
[2021-06-17T15:54:36.172Z] go: downloading sigs.k8s.io/yaml v1.2.0
[2021-06-17T15:54:36.172Z] go: downloading go.mongodb.org/mongo-driver v1.0.3
[2021-06-17T15:54:36.172Z] go: downloading github.com/ghodss/yaml v1.0.0
[2021-06-17T15:54:36.172Z] go: downloading github.com/hashicorp/golang-lru v0.5.1
[2021-06-17T15:54:36.432Z] go: downloading github.com/asaskevich/govalidator v0.0.0-20200428143746-21a406dcc535
[2021-06-17T15:54:36.432Z] go: downloading k8s.io/utils v0.0.0-20201110183641-67b214c5f920
[2021-06-17T15:54:36.432Z] go: downloading github.com/go-openapi/errors v0.19.2
[2021-06-17T15:54:36.432Z] go: downloading gopkg.in/warnings.v0 v0.1.2
[2021-06-17T15:54:36.432Z] go: downloading golang.org/x/time v0.0.0-20200630173020-3af7569d3a1e
[2021-06-17T15:54:36.432Z] go: downloading github.com/mitchellh/mapstructure v1.1.2
[2021-06-17T15:54:36.432Z] go: downloading github.com/go-openapi/spec v0.19.3
[2021-06-17T15:54:36.432Z] go: downloading github.com/emicklei/go-restful v2.9.5+incompatible
[2021-06-17T15:54:36.691Z] go: downloading github.com/go-openapi/swag v0.19.5
[2021-06-17T15:54:36.692Z] go: downloading github.com/go-openapi/jsonpointer v0.19.3
[2021-06-17T15:54:36.692Z] go: downloading github.com/go-openapi/jsonreference v0.19.3
[2021-06-17T15:54:36.692Z] go: downloading github.com/PuerkitoBio/purell v1.1.1
[2021-06-17T15:54:36.692Z] go: downloading github.com/exponent-io/jsonpath v0.0.0-20151013193312-d6023ce2651d
[2021-06-17T15:54:36.692Z] go: downloading k8s.io/component-base v0.20.1
[2021-06-17T15:54:36.692Z] go: downloading github.com/mitchellh/go-wordwrap v1.0.0
[2021-06-17T15:54:36.692Z] go: downloading github.com/russross/blackfriday v1.5.2
[2021-06-17T15:54:36.692Z] go: downloading github.com/mailru/easyjson v0.7.0
[2021-06-17T15:54:36.692Z] go: downloading github.com/MakeNowJust/heredoc v0.0.0-20170808103936-bb23615498cd
[2021-06-17T15:54:36.692Z] go: downloading github.com/docker/spdystream v0.0.0-20160310174837-449fdfce4d96
[2021-06-17T15:54:36.692Z] go: downloading github.com/PuerkitoBio/urlesc v0.0.0-20170810143723-de5bf2ad4578
[2021-06-17T15:54:36.692Z] go: downloading github.com/moby/term v0.0.0-20200312100748-672ec06f55cd
[2021-06-17T15:54:36.692Z] go: downloading github.com/go-stack/stack v1.8.0
[2021-06-17T15:55:08.987Z] Running in /var/lib/jenkins/workspace/est-manager_integrations_PR-1145/src/github.com/elastic/integrations/packages/mysql
[2021-06-17T15:55:09.298Z] + ../../build/elastic-package check -v
[2021-06-17T15:55:09.298Z] elastic-package has been installed.
[2021-06-17T15:55:09.298Z] 2021/06/17 15:55:09 DEBUG Enable verbose logging
[2021-06-17T15:55:09.298Z] Format the package
[2021-06-17T15:55:09.298Z] Error: checking package failed: formatting the integration failed (path: /var/lib/jenkins/workspace/est-manager_integrations_PR-1145/src/github.com/elastic/integrations/packages/mysql, failFast: true): walking through the integration files failed: formatting file failed (path: /var/lib/jenkins/workspace/est-manager_integrations_PR-1145/src/github.com/elastic/integrations/packages/mysql/data_stream/slowlog/_dev/test/pipeline/test-common-config.yml): file is not formatted (path: /var/lib/jenkins/workspace/est-manager_integrations_PR-1145/src/github.com/elastic/integrations/packages/mysql/data_stream/slowlog/_dev/test/pipeline/test-common-config.yml)
[2021-06-17T15:55:09.367Z] Running in /var/lib/jenkins/workspace/est-manager_integrations_PR-1145/src/github.com/elastic/integrations
[2021-06-17T15:55:09.405Z] Archiving artifacts
[2021-06-17T15:55:09.727Z] ‘build/test-results/*.xml’ doesn’t match anything: ‘build’ exists but not ‘build/test-results/*.xml’
[2021-06-17T15:55:09.727Z] No artifacts found that match the file pattern "build/test-results/*.xml". Configuration error?
[2021-06-17T15:55:09.747Z] Recording test results
[2021-06-17T15:55:09.963Z] None of the test reports contained any result
[2021-06-17T15:55:09.963Z] [Checks API] No suitable checks publisher found.
[2021-06-17T15:55:10.266Z] + build/elastic-package stack dump -v --output build/elastic-stack-dump/latest/mysql
[2021-06-17T15:55:10.266Z] 2021/06/17 15:55:10  WARN CommitHash is undefined, in both /var/lib/jenkins/workspace/est-manager_integrations_PR-1145/.elastic-package/version and the compiled binary, config may be out of date.
[2021-06-17T15:55:10.266Z] 2021/06/17 15:55:10 DEBUG Enable verbose logging
[2021-06-17T15:55:10.266Z] 2021/06/17 15:55:10 DEBUG Dump Elastic stack data
[2021-06-17T15:55:10.266Z] 2021/06/17 15:55:10 DEBUG Dump stack logs
[2021-06-17T15:55:10.266Z] 2021/06/17 15:55:10 DEBUG Recreate the output location (path: build/elastic-stack-dump/latest/mysql)
[2021-06-17T15:55:10.266Z] 2021/06/17 15:55:10 DEBUG Dump stack logs for elasticsearch
[2021-06-17T15:55:10.266Z] 2021/06/17 15:55:10 DEBUG running command: /usr/local/bin/docker-compose -f /var/lib/jenkins/workspace/est-manager_integrations_PR-1145/.elastic-package/profiles/default/stack/snapshot.yml -p elastic-package-stack logs elasticsearch
[2021-06-17T15:55:11.209Z] 2021/06/17 15:55:10 DEBUG Dump stack logs for elastic-agent
[2021-06-17T15:55:11.209Z] 2021/06/17 15:55:10 DEBUG running command: /usr/local/bin/docker-compose -f /var/lib/jenkins/workspace/est-manager_integrations_PR-1145/.elastic-package/profiles/default/stack/snapshot.yml -p elastic-package-stack logs elastic-agent
[2021-06-17T15:55:11.780Z] 2021/06/17 15:55:11 DEBUG Dump stack logs for fleet-server
[2021-06-17T15:55:11.781Z] 2021/06/17 15:55:11 DEBUG running command: /usr/local/bin/docker-compose -f /var/lib/jenkins/workspace/est-manager_integrations_PR-1145/.elastic-package/profiles/default/stack/snapshot.yml -p elastic-package-stack logs fleet-server
[2021-06-17T15:55:12.721Z] 2021/06/17 15:55:12 DEBUG Dump stack logs for kibana
[2021-06-17T15:55:12.721Z] 2021/06/17 15:55:12 DEBUG running command: /usr/local/bin/docker-compose -f /var/lib/jenkins/workspace/est-manager_integrations_PR-1145/.elastic-package/profiles/default/stack/snapshot.yml -p elastic-package-stack logs kibana
[2021-06-17T15:55:13.291Z] 2021/06/17 15:55:13 DEBUG Dump stack logs for package-registry
[2021-06-17T15:55:13.291Z] 2021/06/17 15:55:13 DEBUG running command: /usr/local/bin/docker-compose -f /var/lib/jenkins/workspace/est-manager_integrations_PR-1145/.elastic-package/profiles/default/stack/snapshot.yml -p elastic-package-stack logs package-registry
[2021-06-17T15:55:14.232Z] Path to stack dump: build/elastic-stack-dump/latest/mysql
[2021-06-17T15:55:14.232Z] Done
[2021-06-17T15:55:14.275Z] Archiving artifacts
[2021-06-17T15:55:14.722Z] + build/elastic-package stack down -v
[2021-06-17T15:55:14.722Z] 2021/06/17 15:55:14  WARN CommitHash is undefined, in both /var/lib/jenkins/workspace/est-manager_integrations_PR-1145/.elastic-package/version and the compiled binary, config may be out of date.
[2021-06-17T15:55:14.722Z] 2021/06/17 15:55:14 DEBUG Enable verbose logging
[2021-06-17T15:55:14.722Z] Take down the Elastic stack
[2021-06-17T15:55:14.722Z] 2021/06/17 15:55:14 DEBUG running command: /usr/local/bin/docker-compose -f /var/lib/jenkins/workspace/est-manager_integrations_PR-1145/.elastic-package/profiles/default/stack/snapshot.yml -p elastic-package-stack down
[2021-06-17T15:55:15.292Z] The ELASTICSEARCH_IMAGE_REF variable is not set. Defaulting to a blank string.
[2021-06-17T15:55:15.292Z] The KIBANA_IMAGE_REF variable is not set. Defaulting to a blank string.
[2021-06-17T15:55:15.292Z] The STACK_PATH variable is not set. Defaulting to a blank string.
[2021-06-17T15:55:15.292Z] The PROFILE_NAME variable is not set. Defaulting to a blank string.
[2021-06-17T15:55:15.292Z] The ELASTIC_AGENT_IMAGE_REF variable is not set. Defaulting to a blank string.
[2021-06-17T15:55:15.292Z] Removing network elastic-package-stack_default
[2021-06-17T15:55:15.292Z] Network elastic-package-stack_default not found.
[2021-06-17T15:55:15.553Z] Done
[2021-06-17T15:55:15.680Z] Failed in branch mysql
[2021-06-17T15:55:15.726Z] Terminated
[2021-06-17T15:55:15.745Z] Terminated
[2021-06-17T15:55:16.985Z] Running on worker-395930 in /var/lib/jenkins/workspace/est-manager_integrations_PR-1145
[2021-06-17T15:55:17.110Z] [INFO] getVaultSecret: Getting secrets
[2021-06-17T15:55:17.154Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2021-06-17T15:55:19.258Z] + chmod 755 generate-build-data.sh
[2021-06-17T15:55:19.258Z] + ./generate-build-data.sh https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Ingest-manager/integrations/PR-1145/ https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Ingest-manager/integrations/PR-1145/runs/1 FAILURE 420275
[2021-06-17T15:55:19.258Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Ingest-manager/integrations/PR-1145/runs/1/steps/?limit=10000 -o steps-info.json
[2021-06-17T15:55:19.961Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Ingest-manager/integrations/PR-1145/runs/1/tests/?status=FAILED -o tests-errors.json
[2021-06-17T15:55:20.668Z] Retry 1/3 exited 22, retrying in 1 seconds...
[2021-06-17T15:55:22.970Z] Retry 2/3 exited 22, retrying in 2 seconds...
[2021-06-17T15:55:25.239Z] Retry 3/3 exited 22, no more retries left.
[2021-06-17T15:55:25.239Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Ingest-manager/integrations/PR-1145/runs/1/log/ -o pipeline-log.txt

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
Pinging @elastic/integrations (Team:Integrations)

jsoriano commented:
https://github.com/elastic/elastic-package/pull/396 is needed to handle the dropped events in pipeline tests.

mtojek commented on packages/mysql/data_stream/slowlog/elasticsearch/ingest_pipeline/default.json:
```diff
@@ -1,10 +1,19 @@
   "description": "Pipeline for parsing MySQL slow logs.",
   "processors": [{
+    "drop": {
+      "if": "ctx.message ==~ /^# Time:.*/ || ctx.message ==~ /^[\\/\\w\\.]+, Version: .*\\. started with:.*/ms"
```
nit: not just any comment (`# Something`)?

jsoriano replied:
Most of lines in these logs start with `#`, we only drop this one. I think that this time is redundant with other timestamp that is parsed in this module.

mtojek commented on packages/mysql/data_stream/slowlog/manifest.yml:
```diff
@@ -1,16 +1,18 @@
           - /var/log/mysql/mysql-slow.log*
-    title: MySQL slowlog logs
-    description: Collect MySQL slowlog logs
+          - /var/lib/mysql/{{.builtin.hostname}}-slow.log
```
nit: did you check that this is supported correctly (`{{.builtin.hostname}}`)?

jsoriano replied:
Umm, I didn't see this one after reimporting, I will give it a try, good catch.

jsoriano replied:
Trying this I found some issues and ended up adding the docker scenario and system tests :sweat_smile: (WIP, they don't pass yet).

But I am not still sure if this is supported, in any case I am going to use `/var/lib/mysql/*-slow.log*`, that should work the same for most if not all the cases, and we don't need to support these placeholders. This is the only module using them.

mtojek reviewed (approved):
LGTM unless the CI complains.

jsoriano commented:
Moving back to draft till I check what happens with system tests.

mtojek reviewed (approved):
LGTM

Big plus for adding system tests. Nice!

jsoriano commented:
System tests seem to be flaky, taking a look...

jsoriano commented:
Ok, tests should be less flaky now. Things I have changed for that:
* Permissions for log files are managed explicitly in the entrypoint. Not sure if this is the best way to do it, but it is reliable.
* Added a service to wait for mysql to be healthy, this should avoid the errors collecting metrics.

@mtojek as follow ups I think that we should explore a couple of things:
* It seems that the filebeat registry persists between tests, not sure if it is because of this, but I have seen that when running the tests locally, eventually no logs are collected and I have to recreate the stack. It happens in this case with the error data stream but not with the slowlog one, the error one always collects the same file, while the slowlog one depends on the hostname. This makes me think that we should clean the registry between tests.
* For metrics, we are always want to be sure that the service is healthy, if not errors may happen, what make the tests fail. I think this should be managed by the service deployer, if not we are going to need to add a dummy service to all the docker compose files.

mtojek commented:
Thanks for investigation. Both issues are good candidates for improvement. Feel free to create issues in elastic-package and we can discuss them there.

> It seems that the filebeat registry persists between tests, not sure if it is because of this, but I have seen that when running the tests locally, eventually no logs are collected and I have to recreate the stack. It happens in this case with the error data stream but not with the slowlog one, the error one always collects the same file, while the slowlog one depends on the hostname. This makes me think that we should clean the registry between tests.

This one is mysterious to me as different test files (different filenames) are used. We have never observed issues with filebeat registry. I always suggest to depend on live logs instead of copied files. 
I know that there might be a problem when you recreate the `~/.elastic-package/tmp/service_logs` directory as the mounted path is lost and the agent can't see new location. I'm fine with investigating it further.

mtojek commented on packages/mysql/_dev/deploy/docker/entrypoint.sh:
```diff
@@ -1,6 +1,14 @@
 #!/bin/bash
 
+# mysqld creates log files without read permissions for others,
```
Nice hack!

jsoriano commented:
Follow up issues created: https://github.com/elastic/elastic-package/issues/405, https://github.com/elastic/elastic-package/issues/406.
